dev/PagerControl/PagerControl.cpp

```diff
@@ -0,0 +1,773 @@
+        }
+        else if (property == ButtonPanelAlwaysShowFirstLastPageIndexProperty())
+        {
+            UpdateNumberPanel(NumberOfPages());
+        }
+    }
+}
+
+winrt::AutomationPeer PagerControl::OnCreateAutomationPeer()
+{
+    return winrt::make<PagerControlAutomationPeer>(*this);
+}
+
+/* Property changed handlers */
+void PagerControl::OnDisplayModeChanged()
+{
+    // Cache for performance reasons
+    const auto displayMode = DisplayMode();
+
+    if (displayMode == winrt::PagerControlDisplayMode::ButtonPanel)
+    {
+        winrt::VisualStateManager::GoToState(*this, c_numberPanelVisibleVisualState, false);
+    }
+    else if (displayMode == winrt::PagerControlDisplayMode::ComboBox)
+    {
+        winrt::VisualStateManager::GoToState(*this, c_comboBoxVisibleVisualState, false);
+    }
+    else if (displayMode == winrt::PagerControlDisplayMode::NumberBox)
+    {
+        winrt::VisualStateManager::GoToState(*this, c_numberBoxVisibleVisualState, false);
+    }
+    else
+    {
+        UpdateDisplayModeAutoState();
+    }
+}
+
+void PagerControl::UpdateDisplayModeAutoState()
+{
+    const int numberOfPages = NumberOfPages();
+    if (numberOfPages > -1)
+    {
+        winrt::VisualStateManager::GoToState(*this, numberOfPages < c_AutoDisplayModeNumberOfPagesThreshold ?
+            c_comboBoxVisibleVisualState : c_numberBoxVisibleVisualState, false);
+    }
+    else
+    {
+        winrt::VisualStateManager::GoToState(*this, c_numberBoxVisibleVisualState, false);
+    }
+}
+
+void PagerControl::OnNumberOfPagesChanged(const int oldValue)
+{
+    m_lastNumberOfPagesCount = oldValue;
+    const int numberOfPages = NumberOfPages();
+    if (numberOfPages < SelectedPageIndex() && numberOfPages > -1)
+    {
+        SelectedPageIndex(numberOfPages - 1);
+    }
+    UpdateTemplateSettingElementLists();
+    if (DisplayMode() == winrt::PagerControlDisplayMode::Auto)
+    {
+        UpdateDisplayModeAutoState();
+    }
+    if (numberOfPages > -1)
+    {
+        winrt::VisualStateManager::GoToState(*this, c_finiteItemsModeState, false);
+        if (const auto numberBox = m_numberBox.get())
+        {
+            numberBox.Maximum(numberOfPages);
+        }
+    }
+    else
+    {
+        winrt::VisualStateManager::GoToState(*this, c_infiniteItemsModeState, false);
+        if (const auto numberBox = m_numberBox.get())
+        {
+            numberBox.Maximum(INFINITY);
+        }
+    }
+    UpdateOnEdgeButtonVisualStates();
+}
+
+void PagerControl::OnSelectedPageIndexChange(const int oldValue)
+{
+    // If we don't have any pages, there is nothing we should do.
+    // Ensure that SelectedPageIndex will end up in the valid range of values
+    if (SelectedPageIndex() > NumberOfPages() - 1 && NumberOfPages() > -1)
+    {
+        SelectedPageIndex(NumberOfPages() - 1);
+    }
+    else if (SelectedPageIndex() < 0)
+    {
+        SelectedPageIndex(0);
+    }
```

Comment:
we have the ability to coerce dependency properties between values, something like this [https://github.com/microsoft/microsoft-ui-xaml/blob/5d1d3d5ab35cdf408b5ad1426ef28a40dc2b7bf6/dev/Materials/Acrylic/AcrylicBrush.idl#L31](https://github.com/microsoft/microsoft-ui-xaml/blob/5d1d3d5ab35cdf408b5ad1426ef28a40dc2b7bf6/dev/Materials/Acrylic/AcrylicBrush.idl#L31)  should we do this instead of this block of code?

Reply:
Is there a benefit to doing that? After all we would need to run the same checks anyway.